### PR TITLE
Add stronger guarantees to System.gc()

### PR DESCRIPTION
**Pull request by tajila**
The JCL code expects that when System.GC() is called a GC occurs and finalizable objects are finalized. Historically, J9 has never provided this guarantee. We have often had to modify tests to ensure a GC is triggered at the appropriate time. Increasingly, we are finding that more and more tests rely on this behaviour, and most importantly, core JCL code now relies on this behaviour. This PR ensures that a call to System.gc() triggers a GC and runs finalization.

### CODE REVIEW

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Review by gacholio — APPROVED** [review]

**Comment by gacholio** [comment]
Finalization has been deprecated for many releases - it seems odd that the JCL would rely on unspecified and completely unreliable behaviour.

**Review by dmitripivkine — APPROVED** [review]

**Comment by tajila** [comment]
jenkins compile win jdk8

**Comment by tajila** [comment]
@gacholio I had to add an explicit `internalReleaseVMAccess` because finalization cannot be run when VMAccess is held

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Review by gacholio — CHANGES_REQUESTED** [review]
This is no longer valid. You will need to reacquire VM access and exit the VM properly.

**Comment by gacholio** [comment]
`Java_java_lang_ref_Finalizer_runFinalizationImpl` appears to make the same incorrect assumption about VM access and enter/exit - consider fixing this here as well.

**Comment by gacholio** [comment]
Also note the ifdefs in the native - those should probably be mirrored in the JVM call.

**Comment by dmitripivkine** [comment]
> Also note that this doesn't really fix anything - the finalizers themselves run on another thread, so there's no guarantee that any finalization has taken place when the JVM function / native method returns.
> 
> @dmitripivkine Please correct me if `runFinalization` does in fact ensure that finalizers have completed before returning.

 The idea of `runFinalization` is to force Finalizer threads run non-stop until queue is empty.

**Comment by gacholio** [comment]
> The idea of `runFinalization` is to force Finalizer threads run non-stop until queue is empty.

Does this actually mean that all existing finalizers, and any that are created during the call will be processed? The queue would technically be empty while the finalizer was processing the last object in the queue.

**Comment by dmitripivkine** [comment]
> > The idea of `runFinalization` is to force Finalizer threads run non-stop until queue is empty.
> 
> Does this actually mean that all existing finalizers, and any that are created during the call will be processed? The queue would technically be empty while the finalizer was processing the last object in the queue.

I see your point, I will check.

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Review by gacholio — APPROVED** [review]
Regardless of Dmitri's investigation, this is the best we can do (if there are any edge cases they'll be in the GC code). The VM access changes are not strictly necessary, in that atomic-free is currently tolerant of mistakes, but I'd rather see the right thing done, particularly in these cases where the guts of the work are potentially large.

**Comment by tajila** [comment]
The failure was 

```
22:56:31.330 0x11200j9vmutil(j9vm).15     *   ** ASSERTION FAILED ** at /home/jenkins/workspace/Build_JDK21_aarch64_linux_Personal/openj9/runtime/util/mthutil.c:600: ((((uintptr_t)-1) != methodIndex))
```

File: runtime/sunvmi/sunvmi.c [hunk 1]
```diff
@@ -612,13 +612,30 @@ getClassContextIterator(J9VMThread * currentThread, J9StackWalkState * walkState
...
-	VM.javaVM->memoryManagerFunctions->j9gc_modron_global_collect(currentThread);
-	VM.javaVM->internalVMFunctions->internalExitVMToJNI(currentThread);
+#if defined(J9VM_INTERP_ATOMIC_FREE_JNI)
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+#else /* J9VM_INTERP_ATOMIC_FREE_JNI */
+	vmFuncs->internalAcquireVMAccess(currentThread);
+#endif /* J9VM_INTERP_ATOMIC_FREE_JNI */
+
+	mmFuncs->j9gc_modron_global_collect(currentThread);
+	mmFuncs->j9gc_modron_global_collect(currentThread);
```
Comment (keithc-ca) [inline runtime/sunvmi/sunvmi.c:628]: Perhaps a comment is warranted here explaining why `j9gc_modron_global_collect()` is run twice?

**Comment by pshipton** [comment]
Created a doc issue https://github.com/eclipse-openj9/openj9-docs/issues/1161
